docroot/sites/registrar.uiowa.edu/modules/registrar_core/src/Controller/AcademicCalendarController.php

```diff
@@ -0,0 +1,291 @@
+   *   The cache backend service.
+   * @param \Drupal\Core\Render\RendererInterface $renderer
+   *   The renderer.
+   */
+  public function __construct(MauiApi $maui, CacheBackendInterface $cache_backend, RendererInterface $renderer) {
+    $this->maui = $maui;
+    $this->cacheBackend = $cache_backend;
+    $this->renderer = $renderer;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('uiowa_maui.api'),
+      $container->get('cache.default'),
+      $container->get('renderer')
+    );
+  }
+
+  /**
+   * Retrieves calendar data.
+   *
+   * @param \Symfony\Component\HttpFoundation\Request $request
+   *   The current request.
+   *
+   * @return \Symfony\Component\HttpFoundation\JsonResponse
+   *   A JSON response containing the calendar data.
+   */
+  public function getCalendarData(Request $request) {
+    $start = $request->query->get('start');
+    $end = $request->query->get('end');
+    $category = $request->query->all()['category'] ?? [];
+    $subsession = $request->query->get('subsession', '0');
+    $steps = $request->query->get('steps', 0);
+
+    // Ensure category is always an array.
+    if (!is_array($category)) {
+      $category = [$category];
+    }
+
+    // Convert subsession to boolean.
+    $subsession = filter_var($subsession, FILTER_VALIDATE_BOOLEAN);
+
+    $cid = 'registrar_core:academic_calendar:' . $start . ':' . $end . ':' . implode(',', $category) . ':' . ($subsession ? '1' : '0') . ':' . $steps;
+
+    if ($cache = $this->cacheBackend->get($cid)) {
+      $data = $cache->data;
+    }
+    else {
+      $data = $this->fetchAndProcessCalendarData($start, $end, $category, $subsession, $steps);
+      // Cache for 24 hours.
+      $this->cacheBackend->set($cid, $data, time() + 86400);
+    }
+
+    return new JsonResponse($data);
+  }
+
+  /**
+   * Fetches and processes calendar data.
+   *
+   * @param string $start
+   *   The start date.
+   * @param string $end
+   *   The end date.
+   * @param array $categories
+   *   The categories to filter by.
+   * @param bool $subsession
+   *   Whether to include subsessions.
+   * @param int $steps
+   *   The number of sessions to fetch.
+   *
+   * @return array
+   *   The processed calendar data.
+   */
+  private function fetchAndProcessCalendarData($start, $end, $categories, $subsession, $steps) {
+    $current = $this->maui->getCurrentSession();
+    $sessions = $this->maui->getSessionsRange($current->id, max(1, $steps));
+
+    $events = [];
+
+    foreach ($sessions as $session_index => $session) {
+      // Modify the searchSessionDates call to include filtering.
+      $dates = $this->maui->searchSessionDates($session->id, [
+        'startDate' => $start,
+        'endDate' => $end,
+      ], TRUE);
+
+      foreach ($dates as $date) {
+        if ($date->reviewed !== TRUE) {
+          continue;
+        }
```

Comment:
Question: Is there ever a reason that we would return results from MAUI that had results that were not reviewed included? If not, I'm wondering if we should move this to the API layer? Otherwise, it might end up getting repeated.

Reply:
Good question, but one I can't answer because I don't know enough about the data/use-cases. Maybe we seek clarity from Josh or wait till the next time?